File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations
 
def solution(k, dungeons):
    dun_cnt_list = []
    # 방문 순서 확인
    for perm in permutations(range(len(dungeons)), len(dungeons)):
        # 피로도, 방문 수 초기화
        temp_k = k
        dun_cnt = 0
        # 던전 방문
        for i in perm:
            need_k, use_k = dungeons[i]
            # 방문 가능한 경우
            if temp_k >= need_k:
                dun_cnt += 1
                temp_k -= use_k
        # 방문한 던전의 수
        dun_cnt_list.append(dun_cnt)
    # 최대 방문 횟수
    return max(dun_cnt_list)

File: 프로그래머스/2/87946. 피로도/피로도.js
 function permutation (arr, selectNumber) {
    const results = [];
    if (selectNumber === 1) return arr.map((el) => [el]); 

    arr.forEach((fixed, index, origin) => {
      const rest = [...origin.slice(0, index), ...origin.slice(index+1)] 
      const permutations = permutation(rest, selectNumber - 1); 
      const attached = permutations.map((el) => [fixed, ...el]); 
      results.push(...attached); 
    });
    return results; 
}

function solution(k, dungeons) {
    const answers = [];
    const len = dungeons.length;
    const arr = [];
    for (let i = 0; i < len; i++) {
        arr.push(i);
    }
    const combies = permutation(arr, arr.length);
    
    for (comb of combies) {
        let count = 0;
        let rest = k;
        for (index of comb) {
            if (rest >= dungeons[index][0]) {
                count += 1;
                rest -= dungeons[index][1];
            }
        }
        answers.push(count)
    }
    return Math.max(...answers);
}